core/payment/src/api/allocations.rs

```diff
@@ -162,12 +163,83 @@ async fn get_allocation(
     }
 }
 
+fn amend_allocation_fields(
+    old_allocation: Allocation,
+    new_allocation: NewAllocation,
+) -> Allocation {
+    Allocation {
+        allocation_id: old_allocation.allocation_id.clone(),
+        address: new_allocation
+            .address
+            .unwrap_or_else(|| old_allocation.address.clone()),
+        payment_platform: new_allocation
+            .payment_platform
+            .unwrap_or_else(|| old_allocation.payment_platform.clone()),
+        total_amount: new_allocation.total_amount.clone(),
+        spent_amount: old_allocation.spent_amount.clone(),
+        remaining_amount: new_allocation.total_amount - &old_allocation.spent_amount,
+        timestamp: old_allocation.timestamp,
+        timeout: new_allocation.timeout.or(old_allocation.timeout),
+        make_deposit: new_allocation.make_deposit,
+    }
+}
+
 async fn amend_allocation(
     db: Data<DbExecutor>,
     path: Path<params::AllocationId>,
-    body: Json<Allocation>,
+    body: Json<NewAllocation>,
+    id: Identity,
 ) -> HttpResponse {
-    response::not_implemented() // TODO
+    let allocation_id = path.allocation_id.clone();
+    let node_id = id.identity;
+    let new_allocation: NewAllocation = body.into_inner();
+    let dao: AllocationDao = db.as_dao();
+
+    let current_allocation = match dao.get(allocation_id.clone(), node_id).await {
+        Ok(AllocationStatus::Active(allocation)) => allocation,
+        Ok(AllocationStatus::Gone) => {
+            return response::gone(&format!(
+                "Allocation {} has been already released",
+                allocation_id
+            ))
+        }
+        Ok(AllocationStatus::NotFound) => return response::not_found(),
+        Err(e) => return response::server_error(&e),
+    };
+
+    let amended_allocation = amend_allocation_fields(current_allocation.clone(), new_allocation);
+
+    if amended_allocation.remaining_amount < BigDecimal::from(0) {
+        return response::bad_request(
+            &"Amended allocation would be smaller than the amount already spent",
+        );
+    }
+
+    let validate_msg = ValidateAllocation {
+        platform: amended_allocation.payment_platform.clone(),
+        address: amended_allocation.address.clone(),
+        amount: amended_allocation.total_amount.clone() - &current_allocation.total_amount,
+    };
+    match async move { Ok(bus::service(LOCAL_SERVICE).send(validate_msg).await??) }.await {
```

Comment:
Did you check how the validation will behave in corner cases?
For example if we have other allocations, does this function checks the sum of all allocations?
Because if yes, than it already has taken into account the amount from allocation before amendment.

Reply:
The corner case with amending payment-platform no longer exist.

As for validation of allocations, yes, the implementation checks whether the sum of *all* allocations can be paid. Because this takes into account the yet-unmodified allocation, we should only pass the excess for validation (if increasing the allocation) or 0 (if we're shrinking it).

The validation could be skipped entirely for decreasing allocations, but we might wish to extend the validation logic later on so keeping it here makes things easier.